adwords_client/client.py

```diff
@@ -276,24 +275,107 @@ def _batch_operations(self, file_name):
+            'managed_customer': 'ManagedCustomerService',
+            'customer': 'CustomerService',
+            'shared_criterion': 'SharedCriterionService',
+            'campaign_shared_set': 'CampaignSharedSetService',
+            'shared_set': 'SharedSetService',
+            'budget_order': 'BudgetOrderService',
+            'campaign': 'CampaignService',
+            'billing_account': 'BudgetOrderService',
+            'label': 'LabelService',
+            'campaign_ad_schedule': 'CampaignCriterionService',
+            'campaign_targeted_location': 'CampaignCriterionService',
+            'campaign_sitelink': 'CampaignExtensionSettingService',
+            'campaign_callout': 'CampaignExtensionSettingService',
+            'campaign_structured_snippet': 'CampaignExtensionSettingService',
+            'account_label': 'AccountLabelService',
+            'campaign_language': 'CampaignCriterionService'
+        }
+
+        if internal_operation['object_type'] == 'attach_label':
+            if 'ad_id' in internal_operation:
+                raise NotImplementedError()
+            elif 'adgroup_id' in internal_operation:
+                raise NotImplementedError()
+            elif 'campaign_id' in internal_operation:
+                return object_type_service_mapper.get('campaign')
+            elif 'customer_id' in internal_operation:
+                return object_type_service_mapper.get('managed_customer')
+
+        try:
+            return object_type_service_mapper.get(internal_operation['object_type'])
+        except KeyError:
+            logger.debug('There is no custom service class for this object_type: %s', str(internal_operation['object_type']))
+            raise
+
+    def _sync_operations(self):
+        previous_client_id = None
+        previous_service_name = None
+        operation_builder = OperationsBuilder()
+        results = []
+        adw_ops = []
+        for internal_operation in self._read_buffer():
+            client_id = internal_operation['client_id']
+            service_name = self._get_service_from_object_type(internal_operation)
+            for adwords_operation in operation_builder(internal_operation):
+                if client_id != previous_client_id or service_name != previous_service_name:
+                    if previous_client_id is not None and service_name is not None:
+                        service = self.service(previous_service_name)
+                        label_operations = [adw_op for adw_op in adw_ops if 'labelId' in adw_op['operand']]
+                        regular_operations = [adw_op for adw_op in adw_ops if 'labelId' not in adw_op['operand']]
+                        if label_operations:
+                            results.append(service.custom_mutate_labels(previous_client_id, label_operations))
+                        if regular_operations:
+                            results.append(service.custom_mutate(previous_client_id, regular_operations))
```

Comment:
The `custom_` methods were made to insert a `bulk` of operations at once. But the `BaseService` was made thinking that you would be adding operations one-by-one and upload it at the end. There is a better way to do this loop that does not need these methods. Something along the lines of 

```python
        for internal_operation in self._read_buffer():
            client_id = internal_operation['client_id']
            service_name = self._get_service_from_object_type(internal_operation)
            for adwords_operation in operation_builder(internal_operation):
                service = self.service(service_name)
                if previous_client_id and service_name:
                    previous_service = self.service(previous_service_name)
                    if client_id != previous_client_id or service_name != previous_service_name:
                        # time to upload
                        results.extend(previous_service.mutate(previous_client_id))
                        # prepare for next upload
                        previous_service.prepare_mutate()
                else:
                    # this runs on the first loop
                    service.prepare_mutate()

                service.helper.add_operation(adwords_operation)
                previous_client_id = client_id
                previous_service_name = service_name

                # this following "splitting logic" should probably go to your BaseService.ResultProcessor
                # it should make multiple "mutate" calls if necessary and return the results
                # making it transparent to who is adding the operations
                # probably needs creating a new ResultProcessor class or improving on the existing SimpleReturnValue

                # label_operations = [adw_op for adw_op in adw_ops if 'labelId' in adw_op['operand']]
                # regular_operations = [adw_op for adw_op in adw_ops if 'labelId' not in adw_op['operand']]
                # if label_operations:
                #     results.append(service.custom_mutate_labels(previous_client_id, label_operations))
                # if regular_operations:
                #     results.append(service.custom_mutate(previous_client_id, regular_operations))
                # label_operations = []
                # regular_operations = []
                # adw_ops = []
```

Reply:
🤔 I should understand the code better to opine on this one.